### ISSUE TITLE
Convert .dat file based input files to .yaml

### ISSUE THREAD

**Issue by davidrudlstorfer** [imcs-compsim/meshpy#135]
Based on the new option to create input files for 4C based on .yaml files we should also switch the input mechanism in MeshPy to .yaml.

We could even check if the input files are valid with the 4C yaml schemata

I started to work on this - and would drop the .dat file support entirely. There exists a script to convert current .dat files to .yaml and eventually the .dat file support will be dropped in 4C

**Comment by mayrmt** [comment]
Just some context: there is absolutely no timeline and, to my knowledge, no consensus on dropping dat file support. 

I don't say, that meshpy must not switch, but at least based on the facts and not on some plans or ideas for the distant future.

**Comment by isteinbrecher** [comment]
MeshPy will move to YAML! 🚀

A few general thoughts on my side:
- Theoretically, we could use a plain yaml interface, but I think it makes sense to work with [FourCIPP](https://github.com/4C-multiphysics/fourcipp) - especially when we get to the point where we want to read meshes from CubitPy (or any YAML file).
- Functionality that should be dropped while doing this switch (this list will likely get longer):
  - Everything related to parameters/input file sections.
  - `format_replacement` argument in `BoundaryCondition`.
  - Fluid meshes can simply be kept in the yaml dict - no need to handle them specifically.
- Some smaller functionality or behaviour might change during the migration, that is completely fine.
- Some tests might be pointless after the migration -> let's try to identify them if possible
- Currently we read existing dat files and then go through them line by line and when converting each line, we check if we want to just keep the definition or create a "meshpy object". This should be changed to first reading the whole yaml file/dict with FourCIPP and then optionally creating a MeshPy (for the lack of a better word) discretisation. This will drastically simplify the current approach.
- CubitPy migration can be done in parallel, as we can simply use the existing (to be converted to YAML) input files from CubitPy.
- What would be nice, if we have the option to write out the pure mesh information into a separate YAML file -> then we would already have the code structure to ease a potential switch to binary data formats for the mesh.


Open questions:
- How do we want to handle what now is the `InputFile` in the future? I like the idea of having one object that stores the parameters and the mesh. But maybe we don't derive directly from `Mesh` and instead simply have an object with the members (will have better names) `yaml_dict` and `mesh`.
- What is the best way to do this switch while keeping the repo intact and to avoid doing everything at once, also to allow for easy PR review?

I am looking forward to these changes, as I can really see how they will help to clean up some ancient parts of MeshPy.

**Comment by isteinbrecher** [comment]
Tagging @gilrrei

**Comment by isteinbrecher** [comment]
One more thing, I would like to merge #298 before starting this (should be ready within this week). Is there anyone else who wants to finish PRs before the YAML migration? @davidrudlstorfer @eulovi @maxfirmbach @knarfnitram @BishrMaradni @danielwolff1 @rpfsibrain @dharinib98

**Comment by davidrudlstorfer** [comment]
@isteinbrecher thanks for providing your general thoughts - as we've already discussed I've already prepared a rough roadmap (because all our repos now have roadmaps 😉) since the last few weeks

# .dat to .yaml roadmap

## preliminaries

given the recent developments on the 4C and fourcipp side it is time to further plan the transition from the current .dat file based workflow to .yaml.   The present issue should now help to design our vision and plan for the switch. Based on multiple parties involved this should be discussed together in this issue.

It was also mentioned to do an in-person/video meeting for the participants in action (@isteinbrecher, @knarfnitram, @maxfirmbach and me @davidrudlstorfer) but I would prefer to do that once everything is collected here. Otherwise the meeting will just go over the here mentioned points again if people are not up-to-date with everything inside this issue.

## roadmap

For me the following points emerge and should be tackled in the following order:  

1. Convert all .dat file input files from the reference test files to .yaml (with 4C dat2yaml)
2. Introduce fourcipp as an additional dependency of meshpy (like cubitpy)
3. Adjust/update and improve the codebase accordingly
    3.1 Remove the 4C input file functionality entirely from meshpy and start from "scratch". In my opinion we should be able to execute meshpy and all of its function without relying on anything from the 4C folder (see #250). This would be the starting point. Goal is that the 4C functionality is encapsulated within the 4C folder and the mesh can work independently (similar to the Abaqus stuff).
    3.2 Start with a new 4C input file functionality which is based on fourcipp
    3.2 All get_dat_lines functions (spread across the entire code base) should be in once central place, i.e., in the new 4C input file functionality
4. Adjust the test suite so everything passes again
5. Review and iterate the code changes
6. Be happy with the new and improved input/output mechanism 🎉

From my side the following questions arise:
	
- Are the changes in MeshPy and CubitPy interlocked, i.e., do we need to do them in parallel? -> I do not think so
- When this issue is tackled it would be optimal that the development of MeshPy (specifically within the 4C input file is paused). Otherwise the additional effort of rebasing/fixing stuff is not nice. The current main developments are done by @isteinbrecher and me - nevertheless I want to mention it. Therefore, is it possible once the work is tackled to „block“ any changes to related modules in MeshPy (more or less only @isteinbrecher)?  Overall I am very happy to execute this plan. Moreover, I want to do this in the upcoming days/weeks. I am curious on you ideas and input @isteinbrecher @knarfnitram  If we have any questions or need extended functionality (I actually do not think so regarding additional functionality) we can hopefully count on our in-house yaml expert @gilrei

# Q&A from @isteinbrecher 

> - Theoretically, we could use a plain yaml interface, but I think it makes sense to work with [FourCIPP](https://github.com/4C-multiphysics/fourcipp) - especially when we get to the point where we want to read meshes from CubitPy (or any YAML file).

For me that is already set in stone 😅 we must nut implement our own yaml stuff again if we now have one central python interface

> - Functionality that should be dropped while doing this switch (this list will likely get longer):
>   - Everything related to parameters/input file sections.
>   - `format_replacement` argument in `BoundaryCondition`.
>   - Fluid meshes can simply be kept in the yaml dict - no need to handle them specifically.

Yes 🙏🏻

> - Some smaller functionality or behaviour might change during the migration, that is completely fine.

Yes 🙏🏻

> - Some tests might be pointless after the migration -> let's try to identify them if possible

Yes 🙏🏻

> - Currently we read existing dat files and then go through them line by line and when converting each line, we check if we want to just keep the definition or create a "meshpy object". This should be changed to first reading the whole yaml file/dict with FourCIPP and then optionally creating a MeshPy (for the lack of a better word) discretisation. This will drastically simplify the current approach.

Also yes - see my point above ⬆️

> - CubitPy migration can be done in parallel, as we can simply use the existing (to be converted to YAML) input files from CubitPy.

I do not fully understand this point

> - What would be nice, if we have the option to write out the pure mesh information into a separate YAML file -> then we would already have the code structure to ease a potential switch to binary data formats for the mesh.

It's already possible with fourcipp -> you can simply write different sections to different files - therefore, we do not interfere with that from our side

> How do we want to handle what now is the InputFile in the future? I like the idea of having one object that stores the parameters and the mesh. But maybe we don't derive directly from Mesh and instead simply have an object with the members (will have better names) yaml_dict and mesh.

I would completely restructure this and split it up. Otherwise it makes no sense. We should have one abstract class for the input file which simply implements functions like read and write. Then we derive classes from that, i.e., four_c and abaqus, and overload the read and write functions from the base class. The meshpy mesh/or discretization must be uncoupled from the input file

> What is the best way to do this switch while keeping the repo intact and to avoid doing everything at once, also to allow for easy PR review?

These will be very large PR's and i do not think we can simplify a lot. But we could split up everything into two "stages" where the first stage is the removal of the current input file and creation of the new "abstract" input file.

In a second stage we could implement the 4C input file. But this also means that we have no working main between both stages. For me this would be fine


# Conclusing

I am really looking forward to tackling this because this could improve the MeshPy modularity by a lot. Let me know what you think regarding the wall of text @isteinbrecher 😅

**Comment by isteinbrecher** [comment]
Thanks for your thoughts @davidrudlstorfer!


---

> CubitPy migration can be done in parallel, as we can simply use the existing (to be converted to YAML) input files from CubitPy.

I simply meant that we don't have to wait for CubitPy to move to YAML, to work on the MeshPy transition.

> I would completely restructure this and split it up. Otherwise it makes no sense. We should have one abstract class for the input file which simply implements functions like read and write. Then we derive classes from that, i.e., four_c and abaqus, and overload the read and write functions from the base class. The meshpy mesh/or discretization must be uncoupled from the input file

I agree, with the split! I am not sure as of now if it makes sense to provide a "base input file" or we just have one individual input file for each solver we support. Maybe we just have a couple free function in `core` that ease come common mechanism when creating an input file form a MeshPy Mesh. For example `get_coupled_nodes_to_master_map` is used in Abaqus and AceFEM meshes to allow for coupled nodes, but not in 4C because the solver handles coupled nodes differently. But I think we will see that while doing the migration.

> 3.1 Remove the 4C input file functionality entirely from meshpy and start from "scratch". In my opinion we should be able to execute meshpy and all of its function without relying on anything from the 4C folder

100% agree. I think we are close to that, but of course some tweaks are needed (mainly for the point couplings). Slightly related to that, while the base meshpy functionality should be independent of 4C, we can not directly compare meshes, thus we have to create some sort of input file during testing. 

---

What to do next? I think we can directly do your points 1. and 2. and "skip" all the test cases (so we can activate them once they pass and keep on testing that they pass). Then I would simply try to get a first minimal version of the new InputFile working and then add feature after feature. This should help to keep the PRs reviewable (but still not small). Maybe a good point to start is to get the test case `test_mesh_creation_functions_arc_segment_start_end_node` working, for that we only need to port the beam element, the beam material and the nodes.

A sensible plan could be:
1. Do points 1. and 2. and deactivate all test cases
1. Get pure beam mesh working
1. General parameter functionality (this should already cover the CubitPy cases where we don't reconstruct a `Mesh`
1. Node sets and BCs (and couplings?)
1. Generation of a Mesh from CubitPy (or general YAML) file
1. NURBS

I would hope that this allows to break up the changes into sensible chunks. If you agree @davidrudlstorfer, I would open and issue for the first two points so we can talk more about technical details there.

For GitHub development, I have created a branch `migrate_dat_to_yaml` in this repo that we can use to work on the migration, then the `main` branch will stay operational, and once everything on `migrate_dat_to_yaml` works, we can merge the whole thing to main. If we want, we can even activate nightly tests there, to regularly check if the FourCIPP interface works.

---

Regarding the meeting, let's see if we actually need one in the larger group, or maybe just do spontaneous calls if there are questions.

**Comment by davidrudlstorfer** [comment]
@isteinbrecher sounds good

I think we do not need to create issues for all the points above because we are the only people working on it and the above described roadmap sounds good

Thanks for opening the branch - I've just updated it to latest main - so from now on main is "blocked"

I've already started with converting all .dat files to .yaml and this already turns out to be more work than expected 😆 there are many input files which are not compliant with 4C - this is also something bad but I will get it sorted out

**Comment by davidrudlstorfer** [comment]
@isteinbrecher the first point of our roadmap is done with #302 and #303 

Now i would wait until the beginning of next with until this https://github.com/4C-multiphysics/fourcipp/pull/13 is merged. Then we can fully use the functionality of fourcipp and create new input files

**Comment by davidrudlstorfer** [comment]
I've also changed the branch protection rules for `main` in the settings and locked it to not unintentionally merge/push to it.

The reason for that is that you specifically have to change the branch during a PR to our new branch.

**Comment by davidrudlstorfer** [comment]
Collection of ToDo's

- [ ] Change branch protection rule of main (no longer lock it) => at final step when we merge everything to main
- [ ] Update `fourcipp` to pip installable package => once we have pushed `fourcipp` to pypi
- [ ] Increase test coverage again (for 4C tests in workflow file, all other tests in `pyproject.toml` file)
- [ ] Check that we ported/updated all tests => search repo for `Temporarily disabled due to switch to .yaml based input files - check if test is necessary and fix`

**Comment by isteinbrecher** [comment]
Thanks for listing everything! I am fine with not creating issues for all tasks. I will create one though so we can discuss the possible new interface for the `InputFile`, the other things can also be discussed in the respective PRs.

**Comment by davidrudlstorfer** [comment]
@isteinbrecher it looks like with #318 the last step will be done before we can merge m`migrate_dat_to_yaml` to main.

The following ToDo's still need to be done:

- [ ] Change branch protection rule of main (no longer lock it) => at final step when we merge everything to main
- [ ] Use `fourcipp` for the input file
- [ ] Use `fourcipp` for the dictionary comparison
- [ ] Update `fourcipp` to be pulled from pip once it is published to pip
- [ ] Reactivate/fix CubitPy tests once CubitPy is also transitioned to yaml
- [ ] Check & increase minimum testing coverage


Do you have anything in mind?

**Comment by isteinbrecher** [comment]
Thanks for listing the points! I completely agree with all of them, and I am actually really happy with how minor they are - let's see about the switch to fourcipp, but I believe that this should not be remotely as much work as #310. I have some other points that came up during this migration, but they are not directly related to the switch to YAML, so if it is necessary I will open an issue (#311, #309) on some of them.

The only point I can currently think of related to YAML is #308, you could add this to the list as post-fourcipp tasks, then once this switch to fourcipp is done and the compare functions are cleaned up we can discuss if #308 makes sense.

**Comment by davidrudlstorfer** [comment]
Following #310 and #354 the major switch to .yaml is done. I am currently currating all ToDo's from this issue and #354 into a new meta issue for everything

Therefore, closing this
